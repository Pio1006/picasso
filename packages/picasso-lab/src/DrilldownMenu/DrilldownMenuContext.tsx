import { createContext } from 'react'

export interface DrilldownMenuContextProps {
  activeMenuKey?: string
  setActiveMenuKey?: (key?: string) => void
}

export default createContext<DrilldownMenuContextProps>({})
