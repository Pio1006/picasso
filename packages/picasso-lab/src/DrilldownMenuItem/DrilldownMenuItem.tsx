import { ClickAwayListener } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import { Paper } from '@toptal/picasso'
import Popper from '@toptal/picasso/Popper'
import SelectListItem, {
  SelectListItemProps
} from '@toptal/picasso/SelectListItem'
import { OverridableComponent } from '@toptal/picasso-shared'
import React, {
  forwardRef,
  ReactElement,
  useCallback,
  useContext,
  useRef,
  useState
} from 'react'

import styles from './styles'
import DrilldownMenuContext from '../DrilldownMenu/DrilldownMenuContext'

export interface Props extends Omit<SelectListItemProps, 'nested'> {
  /** Nested menu */
  menu?: ReactElement
  /** Container for the the nested drilldown */
  popperContainer?: HTMLElement
}

const generateKey = (() => {
  let count = 0

  return () => String(++count)
})()

const useStyles = makeStyles(styles, {
  name: 'PicassoDrilldown'
})

export const DrilldownMenuItem: OverridableComponent<Props> = forwardRef<
  HTMLElement,
  Props
>(function DrilldownItem (props) {
  const {
    className,
    style,
    selected,
    menu,
    popperContainer,
    onClick,
    ...rest
  } = props
  const classes = useStyles()
  const anchorRef = useRef<HTMLElement>(null)
  const [currentKey] = useState(generateKey)
  const { menuKey, setMenuKey } = useContext(DrilldownMenuContext)
  const opened = currentKey === menuKey

  const handleItemClick = useCallback(
    (event: React.MouseEvent<HTMLElement>) => {
      if (menu && setMenuKey) {
        event.stopPropagation()
        setMenuKey(currentKey)
      }
      if (onClick) {
        onClick(event)
      }
    },
    [currentKey, menu, setMenuKey, onClick]
  )

  const handleAwayClick = useCallback(() => {
    if (setMenuKey) {
      setMenuKey(undefined)
    }
  }, [setMenuKey])

  return (
    <>
      <SelectListItem
        {...rest}
        ref={anchorRef}
        className={className}
        style={style}
        nested={Boolean(menu)}
        selected={opened || selected}
        onClick={handleItemClick}
      />
      {menu && opened && (
        <Popper
          className={classes.popper}
          anchorEl={anchorRef.current}
          placement='right'
          open
          autoWidth={false}
          enableCompactMode
          container={popperContainer}
        >
          <ClickAwayListener onClickAway={handleAwayClick}>
            <Paper className={classes.content}>{menu}</Paper>
          </ClickAwayListener>
        </Popper>
      )}
    </>
  )
})

export default DrilldownMenuItem
