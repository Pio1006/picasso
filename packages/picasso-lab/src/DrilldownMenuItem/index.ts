import { OmitInternalProps } from '@toptal/picasso-shared'

import { Props } from './DrilldownMenuItem'

export { default } from './DrilldownMenuItem'
export type DrilldownItemProps = OmitInternalProps<Props>
