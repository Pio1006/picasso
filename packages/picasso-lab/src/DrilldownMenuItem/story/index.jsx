import { DrilldownMenuItem } from '../DrilldownMenuItem'
import PicassoBook from '~/.storybook/components/PicassoBook'

const componentDocs = PicassoBook.createComponentDocs(
  DrilldownMenuItem,
  'DrilldownMenuItem.Item'
)

export default {
  componentDocs
}
