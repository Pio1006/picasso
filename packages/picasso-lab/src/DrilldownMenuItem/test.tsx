import React, { useCallback, useMemo, useState } from 'react'
import { fireEvent, render, TestingPicasso } from '@toptal/picasso/test-utils'

import DrilldownMenuItem from './DrilldownMenuItem'
import DrilldownMenuContext, {
  DrilldownMenuContextProps
} from '../DrilldownMenu/DrilldownMenuContext'

const TestDrilldownMenuItem = () => {
  const [activeItemKey, setActiveItemKey] = useState<string>()

  const clearActiveItemKey = useCallback(() => setActiveItemKey(undefined), [])

  const contextValue = useMemo(
    (): DrilldownMenuContextProps => ({
      activeItemKey,
      onMouseEnter: setActiveItemKey,
      onClickAway: clearActiveItemKey
    }),
    [activeItemKey, clearActiveItemKey]
  )

  return (
    <TestingPicasso>
      <DrilldownMenuContext.Provider value={contextValue}>
        <DrilldownMenuItem menu={<div data-testid='menu' />} data-testid='item'>
          Item
        </DrilldownMenuItem>
      </DrilldownMenuContext.Provider>
    </TestingPicasso>
  )
}

describe('DrilldownMenuItem', () => {
  it('renders', () => {
    const { container } = render(<TestDrilldownMenuItem />)

    expect(container).toMatchSnapshot()
  })

  it('delegates mouse enter events', () => {
    const { getByTestId, queryByTestId } = render(<TestDrilldownMenuItem />)

    fireEvent.mouseEnter(getByTestId('item'))

    expect(queryByTestId('menu')).toBeInTheDocument()
  })

  it('delegates click away events', () => {
    const { getByTestId, queryByTestId } = render(<TestDrilldownMenuItem />)

    fireEvent.mouseEnter(getByTestId('item'))
    fireEvent.click(document.body)

    expect(queryByTestId('menu')).not.toBeInTheDocument()
  })
})
