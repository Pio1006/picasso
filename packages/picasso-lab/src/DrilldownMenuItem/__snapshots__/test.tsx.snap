// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrilldownMenuItem renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="Picasso-root"
    >
      <li
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
        data-testid="item"
        role="menuitem"
        tabindex="-1"
      >
        <div
          class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
        >
          <div
            class="PicassoContainer-centerAlignItems PicassoContainer-flex"
          >
            <span
              class="PicassoMenuListItem-stringContent"
            >
              Item
            </span>
          </div>
        </div>
      </li>
    </div>
  </div>
</div>
`;
